test/test_charm.py

```diff
@@ -560,27 +560,42 @@ def _on_foo_bar_action(self, event: ops.ActionEvent):
         charm.on.foo_bar_action.emit(id='1')
 
 
-def test_action_event_defer_fails(
-    request: pytest.FixtureRequest, monkeypatch: pytest.MonkeyPatch, fake_script: FakeScript
+@pytest.mark.parametrize(
+    'event,kwargs',
+    [
+        ('start_action', {'id': 2}),
+        ('stop', {}),
+        ('remove', {}),
+        ('secret_expired', {'id': 'secret:123', 'label': None, 'revision': 0}),
+        ('secret_rotate', {'id': 'secret:234', 'label': 'my-secret'}),
+    ],
+)
+def test_inappropriate_event_defer_fails(
+    event: str,
+    kwargs: typing.Dict[str, typing.Any],
+    request: pytest.FixtureRequest,
+    monkeypatch: pytest.MonkeyPatch,
+    fake_script: FakeScript,
 ):
-    cmd_type = 'action'
-
     class MyCharm(ops.CharmBase):
         def __init__(self, framework: ops.Framework):
             super().__init__(framework)
-            framework.observe(self.on.start_action, self._on_start_action)
+            framework.observe(getattr(self.on, event), self._call_defer)
 
-        def _on_start_action(self, event: ops.ActionEvent):
+        def _call_defer(self, event: ops.EventBase):
```

Comment:
I'd prefer a different name for this method, because while we use it to test deferral, in real user code, it would have own semantics, so maybe `def _event_handler` or user event handler or generic event handler?

I wonder if it's worth adding a check that this function was in fact ran, that it's specifcally `event.defer()` call that failed and not something else.

Would it be possible to slap `with pytest.raises(...)` into the body of this method instead?

Reply:
> I'd prefer a different name for this method, because while we use it to test deferral, in real user code, it would have own semantics, so maybe `def _event_handler` or user event handler or generic event handler?

This isn't aiming to look like a real event handler. I think it's better describing what it does than being more generic. I'd use a lambda, avoiding any name, if those could be used as observers.

> I wonder if it's worth adding a check that this function was in fact ran, that it's specifcally `event.defer()` call that failed and not something else.

It seems like that's expanding the scope of this PR somewhat. I guess `RuntimeError` is somewhat generic,s o it's possible that something else in the emitting could cause it, although it's tricky to think of how that might happen without other tests then failing.

My feeling is that this test is checking that calling `defer()` raises in these cases, not that the event handler is properly called, or that calling `defer()` in an event handler works - there are (or should be) other tests for that.

> Would it be possible to slap `with pytest.raises(...)` into the body of this method instead?

That said, I'm not against doing this instead, if people feel like it would be better.

Reply:
As far as the name goes, I think it can be argued both ways (my slight preference is for the "call_defer" name). So let's (ahem) *defer* to Tony's original name.

I think the event handler is small and well-scoped enough that we know it's the defer that failed. I don't think we need to complicate this here.